src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/LimitArrayPoolWriteStream.cs

```diff
@@ -0,0 +1,163 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.SignalR.Internal.Protocol
+{
+    public sealed class LimitArrayPoolWriteStream : Stream
+    {
+        private const int MaxByteArrayLength = 0x7FFFFFC7;
+        private const int InitialLength = 256;
+
+        private readonly int _maxBufferSize;
+        private byte[] _buffer;
+        private int _length;
+
+        public LimitArrayPoolWriteStream() : this(MaxByteArrayLength) { }
```

Comment:
Are we concerned about a max byte array length? (this functionality was already in the stream implementation that I copied this class from)

Reply:
The max is pretty large 😄. We'll need to do some measuring to see how good this is in practice with various message sizes. Pipelines handles write without re-allocating by using a linked list of pooled buffers. Once we have some benchmarks we should compare implementations.